File: Delivery time vs Sorting time.py
# Importing necessary libraries
import pandas as pd # deals with data frame  
import numpy as np  # deals with numerical values

# Input Variable  [x] = Sorting Time
# Output Variable [y]= Delivery Time

del_time = pd.read_csv("C:/Users/SHAJIUDDIN MOHAMMED/Desktop/Data sets/delivery_time.csv")
del_time.columns = "delivery_time", "Sort_time"
del_time.columns

# Exploratory data analysis:
del_time.describe()

#Graphical Representation
import matplotlib.pyplot as plt # mostly used for visualization purposes 

plt.bar(height = del_time.delivery_time, x = np.arange(1,22, 1))
plt.hist(del_time.delivery_time) # Histogram 
plt.boxplot(del_time.delivery_time) # Boxplot

plt.bar(height = del_time.Sort_time, x = np.arange(1, 22, 1))
plt.hist(del_time.Sort_time) # Histogram
plt.boxplot(del_time.Sort_time) # Boxplot

# Scatter plot
plt.scatter(x = del_time['Sort_time'], y = del_time['delivery_time'], color = 'red') 

# Correlation
np.corrcoef(del_time.Sort_time, del_time.delivery_time) 

# Covariance
cov_output = np.cov(del_time.Sort_time, del_time.delivery_time)[0, 1]
cov_output


# Import library
import statsmodels.formula.api as smf

# Simple Linear Regression
model = smf.ols('delivery_time ~ Sort_time', data = del_time).fit()
model.summary()

pred1 = model.predict(pd.DataFrame(del_time['Sort_time']))

# Regression Line
plt.scatter(del_time.Sort_time, del_time.delivery_time)
plt.plot(del_time.Sort_time, pred1, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()

# Error calculation
res1 = del_time.delivery_time - pred1
res_sqr1 = res1 * res1
mse1 = np.mean(res_sqr1)
rmse1 = np.sqrt(mse1)
rmse1

# Model building on Transformed Data(Log Transformation)
# x = log(Sorting Time); y = Delivery Time

plt.scatter(x = np.log(del_time['Sort_time']), y = del_time['delivery_time'], color = 'brown')
np.corrcoef(np.log(del_time.Sort_time), del_time.delivery_time)

model2 = smf.ols('delivery_time ~ np.log(Sort_time)', data = del_time).fit()
model2.summary()

pred2 = model2.predict(pd.DataFrame(del_time['Sort_time']))

# Regression Line
plt.scatter(np.log(del_time.Sort_time), del_time.delivery_time)
plt.plot(np.log(del_time.ST), pred2, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()

# Error calculation
res2 = del_time.delivery_time - pred2
res_sqr2 = res2 * res2
mse2 = np.mean(res_sqr2)
rmse2 = np.sqrt(mse2)
rmse2


# Exponential transformation
# x = Sorting Time; y = log(Delivery Time)

plt.scatter(x = del_time['Sort_time'], y = np.log(del_time['delivery_time']), color = 'orange')
np.corrcoef(del_time.Sort_time, np.log(del_time.delivery_time))

model3 = smf.ols('np.log(delivery_time) ~ Sort_time', data = del_time).fit()
model3.summary()

pred3 = model3.predict(pd.DataFrame(del_time['Sort_time']))
pred3_at = np.exp(pred3)
pred3_at

# Regression Line
plt.scatter(del_time.Sort_time, np.log(del_time.delivery_time))
plt.plot(del_time.Sort_time, pred3, "r")
plt.legend(['Predicted line', 'Observed data'])
plt.show()

# Error calculation
res3 = del_time.delivery_time - pred3_at
res_sqr3 = res3 * res3
mse3 = np.mean(res_sqr3)
rmse3 = np.sqrt(mse3)
rmse3


# Polynomial transformation
# x = Sort_time; x^2 = Sort_time*Sort_time; y = log(delivery_time)

model4 = smf.ols('np.log(delivery_time) ~ Sort_time + I(Sort_time*Sort_time)', data = del_time).fit()
model4.summary()

pred4 = model4.predict(pd.DataFrame(del_time))
pred4_at = np.exp(pred4)
pred4_at

# Regression line
from sklearn.preprocessing import PolynomialFeatures
poly_reg = PolynomialFeatures(degree = 2)
X = del_time.iloc[:, 0:1].values
X_poly = poly_reg.fit_transform(X)


plt.scatter(del_time.Sort_time, np.log(del_time.delivery_time))
plt.plot(X, pred4, color = 'red')
plt.legend(['Predicted line', 'Observed data'])
plt.show()


# Error calculation
res4 = del_time.delivery_time - pred4_at
res_sqr4 = res4 * res4
mse4 = np.mean(res_sqr4)
rmse4 = np.sqrt(mse4)
rmse4


# Choose the best model using RMSE
data = {"MODEL":pd.Series(["SLR", "Log model", "Exp model", "Poly model"]), "RMSE":pd.Series([rmse1, rmse2, rmse3, rmse4])}
table_rmse = pd.DataFrame(data)
table_rmse


# Analysis - Log Model 1 is the best fitted model with Low Rmse

from sklearn.model_selection import train_test_split

train, test = train_test_split(del_time, test_size = 0.2)

finalmodel = smf.ols('delivery_time ~ np.log(Sort_time)', data = train).fit()
finalmodel.summary()


# Predict on test data
test_pred = finalmodel.predict(pd.DataFrame(test))
test_pred

# Model Evaluation on Test data
test_res = test.delivery_time - test_pred
test_sqrs = test_res * test_res
test_mse = np.mean(test_sqrs)
test_rmse = np.sqrt(test_mse)
test_rmse


# Prediction on train data
train_pred = finalmodel.predict(pd.DataFrame(train))
train_pred

# Model Evaluation on train data
train_res = train.delivery_time -  train_pred
train_sqrs = train_res * train_res
train_mse = np.mean(train_sqrs)
train_rmse = np.sqrt(train_mse)
train_rmse

File: Delivery time vs Sorting time.r
# INPUT(X) = Sorting Time
# OUTPUT(Y) = Delivery Time


deltime <- read.csv(file.choose(), header = T)
colnames(deltime) <- c("dt", "sorttime") # dt = delivery time & sorttime = sorting time
View(deltime)

# Exploratory data analysis
summary(deltime)

install.packages("Hmisc")
library(Hmisc)
describe(deltime)
?describe

install.packages("lattice")
library("lattice") 

# Graphical exploration
dotplot(deltime$sorttime, main = "Dot Plot of Sorting Time")
dotplot(deltime$dt, main = "Dot Plot of Delivery Time")

?boxplot
boxplot(deltime$sorttime, col = "dodgerblue4",main = "Sorting Time")
boxplot(deltime$dt, col = "red", horizontal = T, main = "Delivery Time")

hist(deltime$sorttime,main = "Sorting Time")
hist(deltime$dt,main = "Delivery Time")

# Normal QQ plot
qqnorm(deltime$sorttime,main = "Sorting Time")
qqline(deltime$sorttime,main = "Sorting Time")

qqnorm(deltime$dt,main = "Delivery Time")
qqline(deltime$dt,main = "Delivery Time")

hist(deltime$st, prob = TRUE)            # prob=TRUE for probabilities not counts
lines(density(deltime$sorttime))             # add a density estimate with defaults
lines(density(deltime$sorttime, adjust = 2), lty = "dotted")   # add another "smoother" density

hist(deltime$dt, prob = TRUE)            # prob=TRUE for probabilities not counts
lines(density(deltime$dt))             # add a density estimate with defaults
lines(density(deltime$dt, adjust = 2), lty = "dotted")   # add another "smoother" density

# Bivariate analysis
# Scatter plot
plot(deltime$sorttime, deltime$dt, main = "Scatter Plot", col = "Dodgerblue4", 
     col.main = "Dodgerblue4", col.lab = "Dodgerblue4", xlab = "Sorting Time", 
     ylab = "Delivery Time", pch = 20)  # plot(x,y)

attach(deltime)

# Correlation Coefficient
cor(sorttime,dt)

# Linear Regression model
reg <- lm(dt ~ sorttime, data = deltime) # Y ~ X
summary(reg)

confint(reg, level = 0.95)


pred <- predict(reg, interval = "predict")
pred <- as.data.frame(pred)

View(pred)
?predict

# ggplot for adding Regression line for data
library(ggplot2)

ggplot(data = deltime, aes(sorttime, dt) ) +
  geom_point() + stat_smooth(method = lm, formula = y ~ x)


# Evaluation the model for fitness 
cor(pred$fit, deltime$dt)

reg$residuals
rmse <- sqrt(mean(reg$residuals^2))
rmse


# Transformation Techniques(transform the variables to check whether the predicted values are better)
#input = log(x); output = y

plot(log(sorttime), dt)
cor(log(sorttime), dt)

reg_log <- lm(dt ~ log(sorttime), data = deltime)
summary(reg_log)

confint(reg_log,level = 0.95)
pred <- predict(reg_log, interval = "predict")

pred <- as.data.frame(pred)
cor(pred$fit, deltime$dt)

rmse <- sqrt(mean(reg_log$residuals^2))
rmse

# Regression line for data
ggplot(data = deltime, aes(log(sorttime), dt) ) +
  geom_point() + stat_smooth(method = lm, formula = y ~ log(x))


# Log transformation(y)
# input = x; output = log(y)

plot(sorttime, log(dt))
cor(sorttime, log(dt))

reg_log1 <- lm(log(dt) ~ sorttime, data = deltime)
summary(reg_log1)

predlog <- predict(reg_log1, interval = "predict")
predlog <- as.data.frame(predlog)
reg_log1$residuals
sqrt(mean(reg_log1$residuals^2))

pred <- exp(predlog)  # Antilog = Exponential function
pred <- as.data.frame(pred)
cor(pred$fit, deltime$dt)

res_log1 = dt - pred$fit
rmse <- sqrt(mean(res_log1^2))
rmse

# Regression line for data
ggplot(data = deltime, aes(sorttime, log(dt)) ) +
  geom_point() + stat_smooth(method = lm, formula = log(y) ~ x)


# Polynomial models(Non Linear)
# input = x & x^2 (2-degree) and output = log(y)

reg2 <- lm(log(dt) ~ sorttime + I(sorttime*sorttime), data = deltime)
summary(reg2)

predlog <- predict(reg2, interval = "predict")
pred <- exp(predlog)

pred <- as.data.frame(pred)
cor(pred$fit, deltime$dt)

res2 = dt - pred$fit
rmse <- sqrt(mean(res2^2))
rmse

# Regression line for data
ggplot(data = deltime, aes(sorttime, log(dt)) ) +
  geom_point() + stat_smooth(method = lm, formula = y ~ poly(x, 2, raw = TRUE))

# Data Partition(Random Sampling)
n <- nrow(deltime)
n1 <- n * 0.8
n2 <- n - n1

train_ind <- sample(1:n, n1)
train <- deltime[train_ind, ]
test <-  deltime[-train_ind, ]

plot(train$sorttime, train$dt)
plot(test$sorttime, test$dt)

model <- lm(dt ~ log(sorttime), data = train)
summary(model)

confint(model,level=0.95)

log_res <- predict(model,interval = "confidence", newdata = test)
predict_original <- as.data.frame(log_res)
test_error <- test$dt - predict_original$fit # calculate error/residual
test_error

test_rmse <- sqrt(mean(test_error^2))
test_rmse

log_res_train <- predict(model,interval = "confidence", newdata = train)

predict_original_train <- as.data.frame(log_res_train)
train_error <- train$dt - predict_original_train$fit # calculate error/residual
train_error

train_rmse <- sqrt(mean(train_error^2))
train_rmse
